In [1]:
import pandas as pd
import numpy as np

### Q1. Pandas version

In [2]:
pd.__version__

'2.2.2'

## Getting the data

In [3]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv')
df

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                     170            3.0       159.0     3413.433759   
1                     130            5.0        97.0     3149.664934   
2                     170            NaN        78.0     3079.038997   
3                     220            4.0         NaN     2542.392402   
4                     210            1.0       140.0     3460.870990   
...                   ...            ...         ...             ...   
9699                  140            5.0       164.0     2981.107371   
9700                  180            NaN       154.0     2439.525729   
9701                  220            2.0       138.0     2583.471318   
9702                  230            4.0       177.0     2905.527390   
9703                  270            3.0       140.0     2908.043477   

      acceleration  model_year  origin fuel_type         drivetrain  \
0             17.7        2003  Europe  Gasoline    All-wheel drive   
1             17.8        2007     USA  Gasoline  Front-wheel drive   
2             15.1        2018  Europe  Gasoline  Front-wheel drive   
3             20.2        2009     USA    Diesel    All-wheel drive   
4             14.4        2009  Europe  Gasoline    All-wheel drive   
...            ...         ...     ...       ...                ...   
9699          17.3        2013  Europe    Diesel  Front-wheel drive   
9700          15.0        2004     USA  Gasoline    All-wheel drive   
9701          15.1        2008     USA    Diesel    All-wheel drive   
9702          19.4        2011     USA    Diesel  Front-wheel drive   
9703          14.7        2005    Asia    Diesel    All-wheel drive   

      num_doors  fuel_efficiency_mpg  
0           0.0            13.231729  
1           0.0            13.688217  
2           0.0            14.246341  
3           2.0            16.912736  
4           2.0            12.488369  
...         ...                  ...  
9699        NaN            15.101802  
9700        0.0            17.962326  
9701       -1.0            17.186587  
9702        1.0            15.331551  
9703       -1.0            14.884467  

[9704 rows x 11 columns]

# Q2. Records count
How many records are in the dataset?

- [ ] 4704
- [ ] 8704
- [] 9704
- [ ] 17704

In [4]:
df.shape

(9704, 11)

# Q3. Fuel types
What is the most frequent fuel type in the dataset?

In [5]:
df['fuel_type'].unique()

array(['Gasoline', 'Diesel'], dtype=object)

# Q4. Missing values
How many columns in the dataset have missing values?

In [6]:
df.isna().sum()

engine_displacement      0
num_cylinders          482
horsepower             708
vehicle_weight           0
acceleration           930
model_year               0
origin                   0
fuel_type                0
drivetrain               0
num_doors              502
fuel_efficiency_mpg      0
dtype: int64

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9704 entries, 0 to 9703
Data columns (total 11 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   engine_displacement  9704 non-null   int64  
 1   num_cylinders        9222 non-null   float64
 2   horsepower           8996 non-null   float64
 3   vehicle_weight       9704 non-null   float64
 4   acceleration         8774 non-null   float64
 5   model_year           9704 non-null   int64  
 6   origin               9704 non-null   object 
 7   fuel_type            9704 non-null   object 
 8   drivetrain           9704 non-null   object 
 9   num_doors            9202 non-null   float64
 10  fuel_efficiency_mpg  9704 non-null   float64
dtypes: float64(6), int64(2), object(3)
memory usage: 834.1+ KB


In [8]:
df.groupby('fuel_type').count()

engine_displacement  num_cylinders  horsepower  vehicle_weight  \
fuel_type                                                                   
Diesel                    4806           4601        4461            4806   
Gasoline                  4898           4621        4535            4898   

           acceleration  model_year  origin  drivetrain  num_doors  \
fuel_type                                                            
Diesel             4350        4806    4806        4806       4550   
Gasoline           4424        4898    4898        4898       4652   

           fuel_efficiency_mpg  
fuel_type                       
Diesel                    4806  
Gasoline                  4898

# Q5. Max fuel efficiency
What's the maximum fuel efficiency of cars from Asia?

In [9]:
# First, let's look at cars from Asia
asian_cars = df[df['origin'] == 'Asia']
# Now find the maximum fuel efficiency (MPG)
asian_cars['fuel_efficiency_mpg'].max()

23.759122836520497

# Q6. Median value of horsepower
What happens to the median value of horsepower when we fill missing values with the mode?

- [ ] Yes, it increased
- [ ] Yes, it decreased
- [ ] No change

In [10]:
# Step 1: Calculate initial median
print("Initial median of horsepower:", df['horsepower'].median())

# Step 2: Find the most frequent value (mode)
mode_horsepower = df['horsepower'].mode()[0]
print("\nMost frequent horsepower value:", mode_horsepower)

# Step 3: Fill missing values with mode
df['horsepower_filled'] = df['horsepower'].fillna(mode_horsepower)

# Step 4: Calculate new median
new_median = df['horsepower_filled'].median()
print("\nNew median after filling missing values:", new_median)

# Compare the values
print("\nHas the median changed?")
print(f"Original median: {df['horsepower'].median()}")
print(f"New median: {new_median}")
print(f"Difference: {new_median - df['horsepower'].median()}")

Initial median of horsepower: 149.0

Most frequent horsepower value: 152.0

New median after filling missing values: 152.0

Has the median changed?
Original median: 149.0
New median: 152.0
Difference: 3.0


# Q7. Sum of weights
Calculate the sum of weights after performing the specified matrix operations.

In [14]:
# Step 1: Select cars from Asia and required columns
asian_cars = df[df['origin'] == 'Asia']
selected_data = asian_cars[['vehicle_weight', 'model_year']].head(7)

# Step 2: Convert to NumPy array
X = selected_data.values
print("X shape:", X.shape)
print("\nX array:")
print(X)

# Step 3: Compute matrix multiplication of X transpose and X
XTX = X.T @ X
print("\nXTX shape:", XTX.shape)
print("\nXTX matrix:")
print(XTX)

# Step 4: Invert XTX
XTX_inv = np.linalg.inv(XTX)
print("\nXTX_inv matrix:")
print(XTX_inv)

# Step 5: Create array y
y = np.array([1100, 1300, 800, 900, 1000, 1100, 1200])
print("\ny array:")
print(y)

# Step 6: Final calculation - w = (X^T X)^(-1) X^T y
w = XTX_inv @ X.T @ y

print("\nFinal result w:")
print(w)
print("\nSum of all elements in w:", w.sum())

X shape: (7, 2)

X array:
[[2714.21930965 2016.        ]
 [2783.86897424 2010.        ]
 [3582.68736772 2007.        ]
 [2231.8081416  2011.        ]
 [2659.43145076 2016.        ]
 [2844.22753389 2014.        ]
 [3761.99403819 2019.        ]]

XTX shape: (2, 2)

XTX matrix:
[[62248334.33150762 41431216.5073268 ]
 [41431216.5073268  28373339.        ]]

XTX_inv matrix:
[[ 5.71497081e-07 -8.34509443e-07]
 [-8.34509443e-07  1.25380877e-06]]

y array:
[1100 1300  800  900 1000 1100 1200]

Final result w:
[0.01386421 0.5049067 ]

Sum of all elements in w: 0.5187709081074026
